### PR TITLE
support udp protocol

### PR DESCRIPTION
**Pull request by jiafu1115**
This PR try to support udp protocol so that it can fix the issue:  #157 

I just complete the draft without unit-test and without improve the udp socket send logical.
So can you help to take a fast look at to check if overall design is OK without consider UDP socket send part which I will improve it this evening.

### CODE REVIEW

**Comment by jiafu1115** [comment]
Can you help to take a look at overral design to check if it is ok so that I can go ahead with this design.Thanks

**Comment by majst01** [comment]
After a first look this seems to be to many changes just to add udp support (~600 lines changed)
I see your point with the abstraction introduced but i would feel much more comfortable if the first shot of udp support is a smaller change.

**Comment by jiafu1115** [comment]
Thanks for your advice. I will figure out if we can use another style or methods for it so that the changes is smaller. Thus, if we want to support batch send with udp we can't reduce too much code changes.
if only support sync send with udp. it may not involved too many changes.
Whatever, I should complete this PR if the soluction is ok except code change is big.

File: src/main/java/org/influxdb/dto/AbstractBatchPoints.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		 *
+		 * @param tagName
+		 *            the tag name
+		 * @param value
+		 *            the tag value
+		 * @return the Builder instance.
+		 */
+		public T tag(final String tagName, final String value) {
+			this.tags.put(tagName, value);
+			return (T) this;
```
Comment (jiafu1115) [inline src/main/java/org/influxdb/dto/AbstractBatchPoints.java]: if any other method to avoid cast?

**Comment by jiafu1115** [comment]
@majst01 I had completed all code to support UDP protocal. To support batch send with UDP in asyn mode and can't influence old logical of http. I failed to reduce code changes. So can you review the code if you can ignore the code changes number? Thanks.

I notice influx-python and influx-ruby, etc also support UDP, so it is important feature we should support to improve the performance.
To avoid conflict with newest master again and again. Can we take look at this PR ASAP?

Thanks.

**Comment by majst01** [comment]
Still 600 lines changed ?

**Comment by jiafu1115** [comment]
Yes. I compare the implements with python and ruby. The current design is more better than them due to we support async udp and support multi udp port per database. What's more, when enable udp we also can use old api for http and no need to pass extra parameters such as database name which isn't needed by udp.  So it is better design than python though more code changes. Can we review it no matter with code changes number? Thanks

**Comment by jiafu1115** [comment]
@majst01  WDYT? Thanks

**Comment by jiafu1115** [comment]
@majst01  Can you help to enable UDP in influxdb so that I can add tests for this PR.

Change enabled to true
Set udp support database
set bind-address.
etc.
[[udp]]
  enabled = false  
  # bind-address = ""
  # database = "udp"
  # retention-policy = ""

  # These next lines control how batching works. You should have this enabled
  # otherwise you could get dropped metrics or poor performance. Batching
  # will buffer points in memory if you have many coming in.

  # batch-size = 1000 # will flush if this many points get buffered
  # batch-pending = 5 # number of batches that may be pending in memory
  # batch-timeout = "1s" # will flush at least this often even if we haven't hit buffer limit
  # read-buffer = 0 # UDP Read buffer size, 0 means OS default. UDP listener will fail if set above OS max.

  # set the expected UDP payload size; lower values tend to yield better performance, default is max UDP size 65536
  # udp-payload-size = 65536


Many thanks!

**Comment by jiafu1115** [comment]
@majst01 Can you help to enable UDP so that I can test UDP support, Thanks

enable configure can followed this configure:

[[udp]]
  enabled = true
  bind-address = ":8089"
  database = "udp"
  retention-policy = ""

  # These next lines control how batching works. You should have this enabled
  # otherwise you could get dropped metrics or poor performance. Batching
  # will buffer points in memory if you have many coming in.

  batch-size = 1000 # will flush if this many points get buffered
  batch-pending = 5 # number of batches that may be pending in memory
  batch-timeout = "1s" # will flush at least this often even if we haven't hit buffer limit
  read-buffer = 0 # UDP Read buffer size, 0 means OS default. UDP listener will fail if set above OS max.

then make you use the changed configure to run such as :
nohup ./influxd  -config youinfluxdbpath/etc/influxdb/influxdb.conf run  &

**Comment by majst01** [comment]
You have to create a influxdb.conf with this settings enabled and pass it as effective configuration to the influxdb docker container and make it open 8089/udp like so:

*from compile-and-test.sh*
```
docker run -d --name influxdb -p 8086:8086 -p 8089:8089/udp -v ${PWD}/influxdb.conf:/etc/influxdb/influxdb.conf influxdb:${version}
```

If your unit-test now point to influxdb:8089/udp they should work

**Comment by jiafu1115** [comment]
so do you mean that it need me to add the configure by myself? Do I have privilege？

**Comment by majst01** [comment]
@jiafu1115 I can help you, give me some time to enhance *compile-and-test.sh* to make influxdb listen on udp as well.

**Comment by jiafu1115** [comment]
@majst01 OK, Thanks. I  try to create another PR for try, you can help to edit the PR. Thanks

**Comment by jiafu1115** [comment]
@majst01 **OK, When you finish docker file support UDP , notify me.** Then I can add test it.
BTW: I may only support UDP step by step to avoid too much code changes. For example, we can support sync send with UDP as first. This will make us more comfortable.

**Comment by majst01** [comment]
Now you can start adding upd related code which can be integration tested against the influxdb container.

**Comment by jiafu1115** [comment]
@majst01  Thanks. Let's support sync mode writing by UDP as first!

**Comment by majst01** [comment]
I' closing this no

**Pull request closed without merging**